### PR TITLE
8340363: Tag-specific default decorators for UnifiedLogging

### PR DESCRIPTION
**Pull request by anton-seoane**
Hi all,

Currently, the Unified Logging framework defaults to three decorators (uptime, level, tags) whenever the user does not specify otherwise through `-Xlog`. This can result in cumbersome input whenever a specific user that relies on a particular tag(s) has some predefined needs. For example, C2 developers rarely need decorations, and having to manually specify this every time results inconvenient.

To address this, this PR enables the possibility of adding tag-specific default decorators to UL. These defaults are in no way overriding user input -- they will only act whenever `-Xlog` has no decorators supplied and there is a positive match with the pre-specified defaults. Such a match is based on the following:

- Inclusion: if `-Xlog:jit+compilation` is provided, a default for `jit` may be applied.
- Specificity: if, for the above line, there is a more specific default for `jit+compilation` the latter shall be applied. Upon equal specificity cases, both defaults will be applied.
- Additionally, defaults may target a specific log level.

Decorators are also associated with an output file, so an output device may only have one set of decorators. For this reason, if different `LogSelection`s trigger defaults, none is to be applied.

In summary, these defaults may be seen as a "tailored" or "flavoured" version of the existing "uptime-level-tags" current defaults.

Please consider this PR, and thanks!

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8340363](https://bugs.openjdk.org/browse/JDK-8340363): Tag-specific default decorators for UnifiedLogging (**Enhancement** - P4) ⚠️ Issue is not open.

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20988/head:pull/20988` \
`$ git checkout pull/20988`

Update a local copy of the PR: \
`$ git checkout pull/20988` \
`$ git pull https://git.openjdk.org/jdk.git pull/20988/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20988`

View PR using the GUI difftool: \
`$ git pr show -t 20988`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20988.diff">https://git.openjdk.org/jdk/pull/20988.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20988#issuecomment-2358040386)

### CODE REVIEW

**Comment by anton-seoane** [comment]
Reason for covered: I'm an Oracle employee

**Comment by robcasloz** [comment]
Labeling the PR as `hotspot-compiler` because it proposes changing the default decorators of `jit+inlining`.

**Review by dholmes-ora — COMMENTED** [review]
A few comments ...

First these are not "user-specified" decorators, but more developer-specified, but more simply they are "custom default decorators".

Second I'm not clear how tag combinations actually work (is this the "specificity" you refer to? I'm not sure what that means.). E.g. is `-Xlog:jit+inlining` treated the same as `-Xlog:inlining+jit`? And what about wildcards? Will `jit*` trigger this `jit+inlining` default?

> Decorators are also associated with an output file

Right this is key design point in UL: decorators are defined per output. But you are now trying to change that so that decorators are associated with tags instead. his seems a significant deviation from the design.

In any case I'm unclear exactly what happens in this PR - if we log to stdout we can have different decorators, but if we log to a real file they are all disabled? Or does any attempt to control decorators for tags going to the same output result in them all being ignored?

Finally, this is really subjective. You'd really need to socialise the actual proposed changes to the defaults independent of any mechanism to allow it.

**Comment by dholmes-ora** [comment]
This affects all hotspot developers using UL so extending coverage:

/label add hotspot
/label add serviceability

**Comment by robcasloz** [comment]
Nice proposal, Antón! This will make it possible to migrate lots of debug/trace-level ad-hoc logging in the compiler code to the UL while preserving its current format (e.g. time decorators are hardly needed when examining the output of `-XX:+TraceLoopOpts`).

Having said this, I find the following behavior unintuitive. If I run:

```
-Xlog:jit*=debug
```

I get the global default decorators, i.e. `uptime,level,tags`, which is what I expected. But if I run:

```
java -Xlog:jit+compilation=debug,jit+inlining=debug,jit+thread=debug
```

I would expect to get the same decorators, but instead I get the default decorators for `jit+inlining`, i.e. none. Is this intentional?

In general, as a HotSpot developer the behavior I would find most natural is to select the union of all decorators for all chosen tags (regardless of whether the decorators for a tag have been chosen actively by the user, specified as default for the tag, or "inherited" from the global default), as in the first option (`-Xlog:jit*=debug`).

File: src/hotspot/share/logging/logDecorators.cpp [hunk 1]
```diff
@@ -24,6 +24,11 @@
 #include "precompiled.hpp"
 #include "logging/logDecorators.hpp"
 #include "runtime/os.hpp"
+#include "logDecorators.hpp"
+
+const LogLevelType AnyLevel = LogLevelType::NotMentioned;
+#define DEFAULT_DECORATORS \
+  DEFAULT_VALUE(mask_from_decorators(NoDecorators), AnyLevel, LOG_TAGS(jit, inlining))
```
Comment (robcasloz) [inline src/hotspot/share/logging/logDecorators.cpp:31]: As a compiler developer, I agree with the choice of no decorators by default for `jit+inlining`. When this is the only tag selected, there isn't much value in the information provided by the global default decorators.

**Comment by jdksjolen** [comment]
> Finally, this is really subjective. You'd really need to socialise the actual proposed changes to the defaults independent of any mechanism to allow it.

By that you mean the `jit+inlining` default, right? That has been socialized among Oracle's C2 developers if I understand correctly, though it hasn't been done for the wider community. The lack of socialising the changes to the wider the community is an oversight on my part.

File: src/hotspot/share/logging/logDecorators.cpp [hunk 2]
```diff
@@ -24,6 +24,11 @@
 #include "precompiled.hpp"
 #include "logging/logDecorators.hpp"
 #include "runtime/os.hpp"
+#include "logDecorators.hpp"
+
+const LogLevelType AnyLevel = LogLevelType::NotMentioned;
+#define DEFAULT_DECORATORS \
```
Comment (jdksjolen) [inline src/hotspot/share/logging/logDecorators.cpp:30]: I think this should also have the default decorators that UL already have. That is, all data about default decorators is gathered here.

**Comment by anton-seoane** [comment]
@dholmes-ora in addition to the socialising comments that Roberto and Johan have already responded, I'll try to clarify the PR a little (I've updated the title as well to make it clearer):
- `Xlog:jit+inlining` is the same as `Xlog:inlining+jit`, and will get the same treatment
- Wildcards have been chosen to be ignored as you could potentially match too many defaults. In the end, these defaults only attempt to offer a "help" to developers using the same specific `-Xlog` option and that don't want to specify the tagset they are interested on every time
- I am not planning on changing the design idea of "decorators associated with output device". This PR enables having defaults for `-Xlog`-selected tagsets, but once the output device is configured we will end up with the same decorators throughout it (does not matter if it is stdout or a real file)

**Comment by anton-seoane** [comment]
@robcasloz Originally I had thought of these defaults "taking over" if there were no defaults for the rest, but I know what you mean and in a way the uptime-level-tags are some implicit defaults that should also be applied. Merging all default specification as suggested by @jdksjolen would automatically enable this behaviour

**Comment by anton-seoane** [comment]
These defaults are not meant to target a specific selected output, so nothing different would occur. With respect to the first question, right now we would not get any defaults applied as there is a "collision" between A+B and C+D.

That was my original idea, where I assumed it might be unwanted to apply all the possible defaults one over another. However, now (a) I don't think we will have that many defaults to drive this to chaos, and (b) as per @robcasloz feedback I believe it would be useful to apply all.

Going back to your question: right now we do not apply any defaults upon that `-Xlog:A+B,C+D`, but I am working on some changes I will push soon that will change this behaviour to "merge" the default decorators for A+B and C+D

**Comment by dholmes-ora** [comment]
> These defaults are not meant to target a specific selected output, so nothing different would occur.

Sorry but that doesn't make sense. When you set the tagset defaults they are output agnostic, but once you set A+B on the command-line then that is associated with a specific output and so the decorators apply to that output.

> I will push soon that will change this behaviour to "merge" the default decorators for A+B and C+D

What does "merge" mean? union? intersection? I can't see how you can come up with rules that will universally make sense here.

**Comment by anton-seoane** [comment]
Oh, I think I misunderstood you there. In case we have some A+B defaults set, and we run something like `java -Xlog:A+B::uptime`, the "runtime"-specified decorators prevail and the defaults are not to be triggered (e.g. in this case we'd have only uptime decorators, as it's been set explicitly).

With merge I mean the union. It's what @robcasloz suggested above and what most people I've talked to feel to understand to be more logical

**Comment by jdksjolen** [comment]
The main point of importance here is that anything user-specified (from `-Xlog` or `jcmd`) will take priority. We should probably add some information regarding this on the `-Xlog:help` page.

**Comment by anton-seoane** [comment]
Upon further comments and consideration, I am keeping this PR on hold and opening up to further discussion via the hotspot-dev mailing list: https://mail.openjdk.org/pipermail/hotspot-dev/2024-September/094810.html

**Comment by anton-seoane** [comment]
After some review and discussion, I am closing this PR and opening a new (simplified) version of this that aligns with the needed use cases in [8341622: Tag-specific disabled default decorators for UnifiedLogging](https://github.com/openjdk/jdk/pull/21383).

**Pull request closed without merging**